FastSimulation/Tracking/plugins/TrajectorySeedProducer.cc

```diff
@@ -272,347 +288,349 @@ const SeedingNode<TrackingLayer>* TrajectorySeedProducer::insertHit(
+    e.put(output);
 }
 
 
+
 bool
-TrajectorySeedProducer::compatibleWithBeamAxis(
-        const GlobalPoint& gpos1, 
+TrajectorySeedProducer::compatibleWithBeamSpot(
+        const GlobalPoint& gpos1,
         const GlobalPoint& gpos2,
         double error,
         bool forward
-    ) const 
+    ) const
 {
 
-    const double x0 = beamspotPosition.X();
-    const double y0 = beamspotPosition.Y();
-    const double z0 = beamspotPosition.Z();
-    
-    if ( !seedCleaning ) 
-    {
-        return true;
-    }
-
     // The hits 1 and 2 positions, in HepLorentzVector's
     XYZTLorentzVector thePos1(gpos1.x(),gpos1.y(),gpos1.z(),0.);
     XYZTLorentzVector thePos2(gpos2.x(),gpos2.y(),gpos2.z(),0.);
-    
-    // Create new particles that pass through the second hit with pT = ptMin 
-    // and charge = +/-1
 
-    // The momentum direction is by default joining the two hits 
+    // Create new particles that pass through the second hit with pT = ptMin
+    // and charge = +/-1
+    // The momentum direction is by default joining the two hits
     XYZTLorentzVector theMom2 = (thePos2-thePos1);
 
     // The corresponding RawParticle, with an (irrelevant) electric charge
     // (The charge is determined in the next step)
     ParticlePropagator myPart(theMom2,thePos2,1.,magneticFieldMap);
 
-    /// Check that the seed is compatible with a track coming from within
-    /// a cylinder of radius originRadius, with a decent pT, and propagate
-    /// to the distance of closest approach, for the appropriate charge
-    bool intersect = myPart.propagateToBeamCylinder(thePos1,originRadius*1.);
-    if ( !intersect ) 
+    /*
+    propagateToBeamCylinder does the following
+       - check there exists a track through the 2 hits and through a
+    cylinder with radius "originRadius" centered around the CMS axis
+       - if such tracks exists, pick the one with maximum pt
+       - track vertex z coordinate is z coordinate of closest approach of
+    track to (x,y) = (0,0)
+    */
+    if (originRadius>0)
     {
-        return false;
+        bool intersect = myPart.propagateToBeamCylinder(thePos1,originRadius*1.);
+        if ( !intersect )
+        {
+            return false;
+        }
     }
 
     // Check if the constraints are satisfied
     // 1. pT at cylinder with radius originRadius
-    if ( myPart.Pt() < originpTMin ) 
+    if ((originpTMin>0) && ( myPart.Pt() < originpTMin ))
     {
         return false;
     }
 
     // 2. Z compatible with beam spot size
-    if ( fabs(myPart.Z()-z0) > originHalfLength ) 
+    double zConstraint = std::max(originHalfLength,beamSpot->sigmaZ()*nSigmaZ);
+    if ((zConstraint>0) && ( fabs(myPart.Z()-beamSpot->position().Z()) > zConstraint ))
     {
         return false;
     }
+    return true;
+}
 
+bool
+TrajectorySeedProducer::compatibleWithPrimaryVertex(
+        const GlobalPoint& gpos1, 
+        const GlobalPoint& gpos2,
+        double error,
+        bool forward
+    ) const 
+{
 
```

Comment:
@shchauha 

add a comment that this code is currently poorly understood and needs clearer comments in the future